plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,1059 @@
+	hypervisors   hypervisorMap
+	flavors       flavorMap
+	servers       serverMap
+	volumes       volumeMap
+	storagePools  storagePoolMap
+	subnets       subnetMap
+	ports         portMap
+	networks      networkMap
+	aggregates    aggregateMap
+	nova_services nova_serviceMap
+	agents        agentMap
+	stacks        stackMap
+	diag          serverDiag
+}
+
+// Description returns a description string of the input plugin and implements
+// the Input interface.
+func (o *OpenStack) Description() string {
+	return "Collects performance metrics from OpenStack services"
+}
+
+// sampleConfig is a sample configuration file entry.
+var sampleConfig = `
+  ## This is the recommended interval to poll.
+  interval = '30m'
+
+  ## The identity endpoint to authenticate against and get the
+  ## service catalog from
+  identity_endpoint = "https://my.openstack.cloud:5000"
+
+  ## The domain to authenticate against when using a V3
+  ## identity endpoint.  Defaults to 'default'
+  domain = "default"
+
+  ## The project to authenticate as
+  project = "admin"
+
+  ## The user to authenticate as, must have admin rights
+  username = "admin"
+
+  ## The user's password to authenticate with
+  password = "Passw0rd"
+
+  ## Services to be enabled
+  #enabled_services = ["stacks","services", "projects", "hypervisors", "flavors", "servers", "volumes", "storage" , "subnets", "ports", "networks", "aggregates", "nova_services", "agents"]
+  enabled_services = ["services", "projects", "hypervisors", "flavors", "networks", "volumes"]
+
+  #Dependencies
+  # | Service | Depends on |
+  # | servers | projects, hypervisors, flavors |
+  # | volumes | projects |
+
+  ## Collect Server Diagnostics
+  server_diagnotics = false
+
+  InsecureSkipVerify = false
+`
+
+// SampleConfig return a sample configuration file for auto-generation and
+// implements the Input interface.
+func (o *OpenStack) SampleConfig() string {
+	return sampleConfig
+}
+
+// Gather gathers resources from the OpenStack API and accumulates metrics.  This
+// implements the Input interface.
+func (o *OpenStack) Gather(acc telegraf.Accumulator) error {
+	// Perform any required set up
+	if err := o.initialize(); err != nil {
+		return err
+	}
+
+	// Gather resources.  Note service harvesting must come first as the other
+	// gatherers are dependant on this information.
+	gatherers := map[string]func() error{
+		"projects":      o.gatherProjects,
+		"hypervisors":   o.gatherHypervisors,
+		"flavors":       o.gatherFlavors,
+		"servers":       o.gatherServers,
+		"volumes":       o.gatherVolumes,
+		"storage pools": o.gatherStoragePools,
+		"subnets":       o.gatherSubnets,
+		"ports":         o.gatherPorts,
+		"networks":      o.gatherNetworks,
+		"aggregates":    o.gatherAggregates,
+		"nova_services": o.gatherNovaServices,
+		"agents":        o.gatherAgents,
+		"stacks":        o.gatherStacks,
+	}
+	for resources, gatherer := range gatherers {
+		for _, i := range o.EnabledServices {
+			if resources == i {
```

Comment:
Wouldn't
```suggestion
	for _, service := range o.EnabledServices {
		gatherer := gatherers[service]
```
be much easier?